File: blurt_cpp_80211/scrambler.cc
#include "scrambler.h"

Scrambler::Scrambler(uint32_t state_) : state(state_) {
}

int Scrambler::next() {
    state = (0xffff & (state << 1)) ^ (1 & ((state >> 3) ^ (state >> 6)));
    return state & 1;
}

void Scrambler::scramble(const bitvector &input, size_t multipleOf, bitvector &output, uint32_t scramblerState) {
    Scrambler sc(scramblerState);
    size_t pad_bits = 6;
    if (multipleOf != 0) {
        size_t stragglers = (input.size() + 6) % multipleOf;
        if (stragglers)
            pad_bits += multipleOf - stragglers;
    }
    output.assign(input.begin(), input.end());
    output.resize(input.size() + pad_bits);
    for (size_t i=0; i<output.size(); i++)
        output[i] = (uint8_t)(output[i] ^ sc.next());
    for (size_t i=input.size(); i<input.size()+6; i++)
        output[i] = 0;
}

File: blurt_py_80211/scrambler.py
#!/usr/bin/env python
import numpy as np

scrambler_y = np.zeros((127,128), np.uint8)
s = np.arange(128, dtype=np.uint8)
for i in range(127):
    s = np.uint16((s << 1) ^ (1 & ((s >> 3) ^ (s >> 6))))
    scrambler_y[i] = s&1

def scrambler(initial):
    y = scrambler_y[:,np.uint8(initial)]
    i = 0
    while True:
        yield y[i]
        i = (i+1)%127

def scramble(input, multipleof, scramblerState=0x7F):
    sc = scrambler_y[:,np.uint8(scramblerState)]
    pad_bits = 6
    if multipleof != None:
        stragglers = (input.size + 6) % multipleof
        if stragglers:
            pad_bits += multipleof - stragglers
    output = np.r_[np.array(input, int)&1, np.zeros(pad_bits, int)]
    output ^= np.tile(sc, (output.size+126)/127)[:output.size]
    output[input.size:input.size+6] = 0
    return output
